README
RDF-Sesame version 0.16
========================

DESCRIPTION

A Perl module to interact with the Sesame RDF server (http://openrdf.org).
For more complete documentation, please see the module's POD.

TESTING

The tests are designed to interact with a working Sesame server.  If you
don't set the environment variables listed below, the tests should
still pass, but will have skipped a great deal, therefore, you
wouldn't know if your installation actually works.

The tests will DESTROY ANY DATA in the repository you provide.  It is
recommended that you specifically create a repository to use with the
tests or use the default mem-rdf-db repository that comes with a default
Sesame installation.  The repository must have inferencing DISABLED.
RDF::Sesame works correctly when repositories have inferencing
enabled, but the tests are written on the assumption that Sesame will
not add extra triples to the store.

If you've made it through all that, here are the secret code words you
need to get the tests working.  You must set the following environment
variables:

   SESAME_URI=http://<user>:<password>@<server>:<port>/<directory>
   SESAME_REPO=<repository>

In the above settings, the items in <> are patterns for which you
should substitute.

   user      : the name of a user with permission to read and write
   password  : the password of the user
   server    : the hostname or IP address of the Sesame server
   port      : the port on which Sesame is listening
   directory : this is almost always 'sesame'
   repository: the Sesame repository to run tests against.  If you're
               using the default, in-memory repository, this would
               be 'mem-rdf-db'

KNOWN ISSUES

There is a problem with the implementation of Native repositories in
Sesame 1.2.0.  The RDF::Sesame tests will fail on a native repository
under that Sesame version. The bug is fixed in Sesame 1.2.1

INSTALLATION

Installation uses Module::Build.  The mantra for installation is as
follows.

perl Build.PL
./Build
./Build test
./Build install

DEPENDENCIES

   Carp            - standard module
   Data::Table
   LWP::UserAgent
   Test::More      - version 0.31 or higher
   URI
   XML::Simple

COPYRIGHT AND LICENCE

Copyright (C) 2005-2006 by Michael Hendricks

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.



Changes
---
version : 0.16
date    : 2007-05-25T15:58:29Z
changes :
    - Documented some of the limits of using the binary results format
    - Correct triple count when uploading more than 999 triples.  Thanks Benno Blumenthal.
    - Add server_file option to upload_uri.  Thanks Benno Blumenthal.
---
version : 0.15
date    : 2006-04-19T19:24:13Z
changes :
    - >
      added support for binary table results format which parses
      *much* (50-500 times) faster than the XML results format."
    - use binary table results format by default
    - minor code cleanup

---
version : 0.14
date    : 2006-04-17T20:46:10Z
changes :
    - updated the year on copyright statements
    - added extract() to RDF::Sesame::Repository
    - added construct() to RDF::Sesame::Repository (rt.cpan.org #13600)

---
version : 0.13
date    : 2006-02-07T09:45:46Z
changes :
    - added test coverage statistics to the POD
    - added compatibility statement for Sesame 1.2.3 and 1.2.4
    - added rudimentary debugging/profiling support (RDFSESAME_DEBUG)
    - corrected the connect() call in the SYNOPSIS
    - added a test for POD coverage
    - add "use strict" and "use warnings" to all modules

---
version : 0.12
date    : 2005-09-29T17:31:30Z
changes : 
    - highlighted that inferencing must be disabled to run tests
    - added compatibility table
    - added a module signature

---
version : 0.11
date    : 2005-07-14T15:58:24Z
changes :
    - corrected the repository.t test

---
version : 0.10
date    : 2005-07-05T22:27:10Z
changes :
    - original version


lib/RDF/Sesame/Response.pm
package RDF::Sesame::Response;

use strict;
use warnings;
use XML::Simple;

our $VERSION = '0.16';

# simple accessors
sub errstr        { return $_[0]->{error}   }
sub http_response { return $_[0]->{http}    }
sub parsed_xml    { return $_[0]->{parsed}  }
sub success       { return $_[0]->{success} }
sub xml           { return $_[0]->{xml}     }

sub content {
    my ($self) = @_;
    return $self->http_response()->content();
}

sub is_xml {
    my ($self) = @_;
    my $content_type = $self->http_response()->header('Content-type');
    return $content_type =~ m{ text/xml }xms;
}

sub is_binary_results {
    my ($self) = @_;
    my $content_type = $self->http_response()->header('Content-type');
    return $content_type =~ m{ application/x-binary-rdf-results-table }xms;
}

#
# Creates a new RDF::Sesame::Response object.
#
# The parameter $response is an HTTP::Response object
#
sub new {
    my ($class, $r) = @_;

    my $self = bless {
        http    => $r, # our original HTTP::Response object
        success => 0,  # was the command sucessful?
        xml     => '', # the XML from the server
        error   => '', # the error message from the server
        parsed  => {}, # a hashref representing the parsed XML
    }, $class;

    # return an empty object if we got no HTTP::Response
    return $self unless $r;

    if ( !$r->is_success() ) {
        $self->{error} = $r->message();
        return $self;
    }

    $self->{success} = 1;
    return $self if !$self->is_xml();

    # because the XML for tuples prevents XML::Simple
    # from retaining the attribute order, we do a transform
    # to improve the XML.
    # See the documentation for _fix_tuple()
    my $xml = $self->{xml} = $r->content();
    $xml =~ s#<tuple>(.*?)</tuple>#_fix_tuple($1)#siegx;

    # TODO call a custom XML::SAX parser instead
    $self->{parsed} = XMLin(
        $xml,
        ForceArray => [
            qw(repository status notification
                columnName tuple  attribute  
                error                         )
        ],
        KeyAttr    => [ ],
    );

    # examine the XML for error responses
    if( exists $self->{parsed}{error} ) {
        $self->{success} = 0;
        $self->{error} = @{$self->{parsed}{error}}[0]->{msg};
    }

    return $self;
}

# The XML returned by Sesame after evaluating a table query
# contains the attributes of each tuple in the same order as
# the attribute (column) names.  This order is not preserved by XML::Simple
# so we need to transform the provided-XML into a more useful form.
# This could be done with XSLT but that's overkill, so
# we jut use regular expressions.
#
# An example transform would take XML like this
#
# <tuple>
#   <bNode>node1</bNode>
#   <literal>Hello</literal>
#   <uri>http://example.com</uri>
#   <literal datatype="http://example.org/string">World!</literal>
# </tuple>
#
# and transform it into XML like this
#
# <tuple>
#   <attribute type='bNode'>node1</attribute>
#   <attribute type='literal'>Hello</attribute>
#   <attribute type='uri'>http://example.com</attribute>
#   <attribute type='literal' datatype="http://example.org/string">World!</attribute>
# </tuple>
sub _fix_tuple {
    my $content = shift;

    $content =~ s#
        <\s*(bNode|literal|uri)(.*?)>(.*?)</\1>
       #<attribute type='$1'$2>$3</attribute>#sgix;

    $content =~ s#
        <\s*(null)(.*?)\s*/\s*>
       #<attribute type='$1'$2 />#sgix;

    return "<tuple>$content</tuple>";
}

1;

__END__

=head1 NAME

RDF::Sesame::Response - A response of a Sesame server to a command

=head1 DESCRIPTION

This class is mostly used internally, but it's documented here in case
others find use for it. This object contains information about the
response from the Sesame server and provides useful ways to access the
information.  After one executes RDF::Sesame::Connection::command or
RDF::Sesame::Repository::command, one should check the response object
to determine if the command was completed successfully.

Just to say it one more time, most users won't need this class.  Use
the methods provided by RDF::Sesame::Repository instead.

=head1 METHODS

=head2 new

Constructs a new RDF::Sesame::Response object from an L<HTTP::Response> object
from a Sesame server.

=head2 content

This method is a shortcut for C<< $r->http_response()->content() >>.  Namely,
it returns the content of the L<HTTP::Response> object from which this
RDF::Sesame::Response object was constructed.

=head2 errstr

Returns the error message provided by the Sesame server.  If there was no
error message, returns the empty string.

=head2 http_response

Returns the HTTP::Response object representing the original HTTP response
from the server.  This allows one access to the lower-level details about
the response.

=head2 is_binary_results

Returns a true value if the Sesame response contained a binary RDF table
results payload; otherwise, returns false.  The determination is based on the
C<Content-Type> header of the HTTP response.

=head2 is_xml

Returns a true value if the Sesame response contained an XML payload;
otherwise, returns false.  The determination is based on the C<Content-Type>
header of the HTTP response.

=head2 parsed_xml

Returns a hashref representation of the XML returned by the Sesame server.
If no XML is available, returns an empty hashref.  XML::Simple is used
to parse the XML.

=head2 success

Returns 1 if the response from the server indicates that the command
was completed successfully, otherwise it returns the empty string.

=head2 xml

Returns the raw XML returned by the Sesame server.  This value will
only be available if the response actually included XML.  If it didn't,
the empty string is returned.

=head1 AUTHOR

Michael Hendricks  <michael@ndrix.org>

=head1 LICENSE AND COPYRIGHT
 
Copyright (c) 2005-2006 Michael Hendricks (<michael@ndrix.org>). All rights
reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
 


lib/RDF/Sesame/Connection.pm
# vim modeline vim600: set foldmethod=marker :
package RDF::Sesame::Connection;

use strict;
use warnings;

use Carp;
use LWP::UserAgent;
use Time::HiRes qw( gettimeofday tv_interval );

use RDF::Sesame;
use RDF::Sesame::Repository;

our $VERSION = '0.16';

=head1 NAME

RDF::Sesame::Connection - A connection to a Sesame server

=head1 DESCRIPTION

This class represents a connection to a specific Sesame server and provides
methods which are generally useful for interacting with a Sesame server.
This class is predominantly used to create an RDF::Sesame::Repository object
which is in turn used to execute queries, remove triples, upload triples, etc.

=head1 METHODS

=head2 open ( %opts )

Creates an RDF::Sesame::Repository object.  There is no communication with
the Sesame server during this call, so it should be very fast.  The C<%opts>
parameter provides a series of named options to use when creating the Repository
object.  A list of options which are currently understood is provided below.
If a single scalar is given instead of C<%opts>, the scalar will be treated
as the value of the 'id' option.

=head3 id

The ID of the repository that you want to open.  The method will return
successfully even if there is no repository on the server with the
repository ID provided.  However, subsequent attempts to do anything
useful with the Repository object will fail.

If you're not sure what the valid repository IDs are, use the
C<repositories> method documented below.  If no repository ID is given,
the method returns the empty string.

=head3 query_language

This is equivalent to calling query_language() on the newly created Repository
object.  See that documentation for further information.

=head3 strip

This option serves as a default for the strip option to
RDF::Sesame::Repository::select  Typically, that method defaults to 'none'
however specifying this option when opening the repository changes the default.

=cut

sub open {
    RDF::Sesame::Repository->new(@_);
}

=head2 repositories ( [ $refresh ] )

When called in list context, returns a list of the IDs of the repositories
which are available through this connection.  The response in scalar context
is reserved for future use.  So don't call repositories() in scalar
context yet (currently, it returns C<undef>).

Only the first call to repositories() communicates with the server.
Subsequent calls return a cached copy of the results from the first
communication.  If you want to get new results directly from the server, 
pass a true value for the $refresh parameter.  These new results will replace
the previously cached results.

=cut

sub repositories {
    my $self = shift;

    return undef unless wantarray;

    # check the repositories cache
    if( $self->{repos} && !$_[0] ) {
        return @{ $self->{repos} };
    }

    # call listRepositories on the server
    my $r = $self->command(undef, 'listRepositories');
    my @repos;
    if( $r->success ) {
        foreach ( @{ $r->parsed_xml->{repository} } ) {
            push(@repos, $_->{id});
        }

        $self->{repos} = \@repos;
    } else {
        # there was an error, so no repositories are available
        return ();
    }

    return @repos;
}

=head2 disconnect

Logs out of the Sesame server and closes any connections to the server.
Attempts to use this object or any of the RDF::Sesame::Repository objects
created by this object after it has been disconnected will result in ugly
problems.

This method is B<not> called when an RDF::Sesame::Connection object is
destroyed.  If you want to explicitly logout of the server, you must
call this method.

Returns true upon success, false otherwise.

=cut

sub disconnect {
    my $self = shift;

    my $resp = 0;
    if( $self->{authed} ) {
        $resp = $self->command(undef, 'logout')->success;
    } else {
        $resp = 1; # we can't fail if we're already logged out
    }

    if( $resp ) {
        $self->{authed} = 0;
        delete $self->{ua};
    }

    return $resp;
}

=head1 INTERNAL METHODS

These methods might be useful to some users, so they're documented, but
most will never need them.

=head2 command ( $repository_id, $name [, $parameters ] )

This method executes a command against a repository
using the parameters provided in the hashref.  The intended way to execute
commands against the server is to use an RDF::Sesame::Repository object.

The C<$repository_id> parameter is just a shortcut for adding a I<repository>
parameter in the call to the server.  If you pass C<undef> as the repository
ID, the method should still work fine and no I<repository> parameter will
be passed to the server.

The result of this method is an RDF::Sesame::Response object.  That object
can be used to determine whether the command succeeded or failed.

 Parameters :
    $repository_id  The ID of the repository you want to execute
        the query against.

    $name  The name of the command to execute.

    $parameters  An optional hashref containing the parameter
        names and values to pass to the server when executing
        the command.
 
 Return : 
    RDF::Sesame::Response

=cut

# TODO use XML::SAX instead of XML::Simple (details follow)
# The basic implementation might be something like this
#
#   my ($self, $cmd) = @_;
#   my $handler_class = $handlers{$cmd};
#   my $handler = $handler_class->new();
#   $parser = XML::SAX::ParserFactory( Handler => $handler );
#   $self->{ua}->post(
#       ...,
#       ':content_cb' => sub { $parser->parse_string(...) }
#   );
#   return $handler->response();
#
# I should be able to implement the above if I make the new response
# objects implement the current Response interface.  Once that works,
# I can change the way the old code uses the response objects
# (if that's still necessary).

{ my $count = 0;
sub command {
    my $self = shift;

    # make sure we have a hash
    my $params;
    if( ref($_[2]) eq 'HASH' ) {
        $params = $_[2];
    } else {
        $params = {};
    }

    # add the repository name to the hash
    $params->{repository} = $_[0] if defined $_[0];

    my $cmd = $_[1];

    # make the request. Either GET or POST depending on the command
    my $cmd_uri = $self->{server} . $cmd;
    my $r;  # the server's HTTP::Response
    my $content_cb = delete $params->{':content_cb'};
    my $start = [ gettimeofday() ];
    if( $cmd eq 'listRepositories' or $cmd eq 'logout' ) {
        # send a request using HTTP-GET
        $r = $self->{ua}->get( $cmd_uri, %$params );
    } else {
        # send a request using HTTP-POST ('multipart/form-data' encoded)
        $r = $self->{ua}->post(
            $cmd_uri,
            {},  # empty form since the real stuff is in 'Content'
            Content_Type => 'form-data',
            Content      => $params,
            ( $content_cb ? (':content_cb' => $content_cb) : () )
        );
    }

    # make an RDF::Sesame::Response object for return
    my $response = RDF::Sesame::Response->new($r);

    if ( $ENV{RDFSESAME_DEBUG} ) {
        my $elapsed = int( 1000 * tv_interval($start) );  # in milliseconds
        printf STDERR "Command %d : Ran $cmd in $elapsed ms\n", $count++;
    }

    return $response;
}
}

# This method should really only be called by
# RDF::Sesame::connect and it's documented there, so
# there's no need to document it here also.

sub new {
    my $class = shift;

    # Establish the defaults for each option
    my %defaults = (
        host      => 'localhost',
        port      => 80,
        directory => 'sesame',
        timeout   => 10,
    );

    my %opts;
    if( @_ == 1 ) {
        $opts{host} = shift;
    } else {
        %opts = @_;
    }

    if( $opts{host} and $opts{host} =~/^(.*):(\d+)$/ ) {
        $opts{host} = $1;
        $opts{port} = $2;
    } elsif( $opts{uri} ) {
        require URI;
        my $uri = URI->new( $opts{uri} );

        # set the individual options based on the URI
        $opts{host} = $uri->host;
        $opts{port} = $uri->port;
        $opts{directory} = $uri->path;

        my($user, $pass) = split(/:/, $uri->userinfo || '', 2);
        $opts{username} = $user if defined $user;
        $opts{password} = $pass if defined $pass;
    }

    # set the defaults
    while( my ($k,$v) = each %defaults ) {
        $opts{$k} = $v unless exists $opts{$k};
    }

    # normalize the sesame directory
    $opts{directory} =~ s#^/+##g;
    $opts{directory} =~ s#/+$##g;

    # create a user agent for making HTTP requests
    my $ua = LWP::UserAgent->new(
            agent      => "rdf-sesame/$RDF::Sesame::VERSION ",
            keep_alive => 1,
            cookie_jar => {},
            timeout    => $opts{timeout},
    );

    # create our new self
    my $self = bless {
        server => "http://$opts{host}:$opts{port}/$opts{directory}/servlets/",
        ua     => $ua,
        authed => 0,     # are we logged in?
        repos  => undef, # list of available repositories
    }, $class;

    # do we even need to login ?
    return $self unless defined $opts{username};

    # yup, so go ahead and do it
    my $r;
    $opts{password} = '' unless defined $opts{password};
    $r = $self->command(
        undef,
        'login',
        {user=>$opts{username}, password=>$opts{password}}
    );

    unless( $r->success ) {
        $RDF::Sesame::errstr = $r->errstr;
        return '';
    }

    $self->{authed} = 1;
    return $self;
}

=head1 AUTHOR

Michael Hendricks <michael@ndrix.com>

=cut

return 1;


lib/RDF/Sesame.pm
package RDF::Sesame;

use strict;
use warnings;

use RDF::Sesame::Connection;
use RDF::Sesame::Response;
use RDF::Sesame::TableResult;

our $VERSION = "0.16";
our $errstr;  # holds the error string from a failed connect

=head1 NAME

RDF::Sesame - Interact with Sesame RDF servers

=head1 SYNOPSIS

 use RDF::Sesame;
 
 # Connect anonymously to Sesame on localhost port 80
 my $sesame = RDF::Sesame->connect;
 
 # or anonymously on a remote Sesame server
 $sesame = RDF::Sesame->connect('openrdf.org');
 
 # or explicitly specify the options
 $sesame = RDF::Sesame->connect(
    host      => "openrdf.org",
    port      => 80
    directory => "sesame",
    username  => "testuser",
    password  => "opensesame"
 ) or die "Couldn't connect to Sesame : $RDF::Sesame::errstr\n";
 
 # or explicitly specify using a single URI
 $sesame = RDF::Sesame->connect(
    uri => 'http://testuser:opensesame@openrdf.org:80/sesame'
 );
 
 my @repos = $sesame->repositories;
 
 # open a repository the easy way
 my $vcard  = $sesame->open("vcard");
 
 # or the more flexible way (allowing other named options)
 my $museum = $sesame->open(id => "museum");
 
 my $serql = <<END;
    select x, given, family
    from
     {x} vCard:N {n},
     {n} vCard:Given  {given};
         vCard:Family {family}
    using namespace
      vCard = <http://www.w3.org/2001/vcard-rdf/3.0#>
 END
 
 my $results = $vcard->select(
     query    => $serql,
     language => "SeRQL"
 );
 
 $vcard->query_language("SeRQL");
 $results = $vcard->select($serql);
 
 # $results is a Data::Table object (via RDF::Sesame::TableResults)

=head1 DESCRIPTION

The RDF::Sesame module implements a wrapper around the RESTful API
(HTTP protocol) provided by Sesame L<http://openrdf.org>.  It facilitates
connecting, adding data and querying a Sesame server from Perl.

RDF::Sesame itself only provides a method for creating a Connection
object.  See the documentation for L<RDF::Sesame::Connection>,
L<RDF::Sesame::Repository>, and L<RDF::Sesame::TableResults>.

=head1 STATUS

This module is in beta testing.  The algorithms have been tested
thoroughly and are stable.  The API might change in a few particulars,
hopefully in a backwards-compatible manner.

=head1 METHODS

=head2 connect ( %opts )

Connects to a Sesame server and creates an instance of an
RDF::Sesame::Connection object.  This object can be used to create
an RDF::Sesame::Repository object or (less likely) to execute commands
against a Sesame server.  Creating an RDF::Sesame::Repository object is
the preferred way, so do that.

The HTTP connection to the server is created with Keep-Alive enabled in
an attempt to make consecutive requests speedier.

The C<%opts> parameter is a hash of options to use when creating the
connection.  Below is a list of the currently understood options.  If a single
scalar is provided to connect(), it will be treated as the host and
all other options will use the default values.

=head3 host

The host name or address of the sesame server. For example 'openrdf.org'
or 'openrdf.org:8080'.  It's B<not> a URI.

Default: localhost

=head3 port

The port number on which the Sesame server is listening.

Default: 80

=head3 directory 

The sesame directory on the host.  The directory should be specified without
leading or trailing '/' characters.  However, if those characters are provided,
they will be stripped before further processing.

Default: sesame

=head3 username

The username to use for logging in to the server.  If this option is not
specified (or C<undef>), no login will be attempted and only publicly available
repositories will be visible.

=head3 password

The password to use for logging in to the server.  If this option is not
supplied but the 'username' option is specified, a password consisting of
the empty string will be used to log in.

=head3 uri

A URI specifying how to connect to the Sesame server.  All of the above
options may be specified at once by providing them in the URI in standard
format.  Here is an example URI

 http://username:password@host:port/directory

If any of the parts are not specified, the defaults explained above will be
used instead.

=head3 timeout

When communicating with the Sesame server, RDF::Sesame usually waits 10
seconds for the server to respond.  If there is no response within that
time, the server is considered unreachable.  This option allows you
to change this behavior.

=cut

sub connect {
    shift @_;  # remove the class name
    RDF::Sesame::Connection->new(@_);
}

=head1 CONFIGURATION AND ENVIRONMENT
 
RDF::Sesame can generate very limited debugging/profiling output.  By setting
the environment variable C<RDFSESAME_DEBUG> to a true value, each request to
the Sesame server will generate a message on STDERR.  Here is a short sample:

   Command 0 : Ran login in 30 ms
   Command 1 : Ran evaluateTableQuery in 45 ms
   Command 2 : Ran evaluateTableQuery in 120 ms

The time value generated with each message is the number of milliseconds it
took to process the command including network time, database processing time
and response processing time.

=head1 COMPATIBILITY

The following table indicates RDF::Sesame's compatibility with different
versions of Sesame and different sail implementations.

  Sesame 1.2.6 : native - OK
                 memory - OK
  Sesame 1.2.4 : native - OK
                 memory - OK
  Sesame 1.2.3 : native - OK
                 memory - OK
  Sesame 1.2.2 : native - OK
                 memory - OK
  Sesame 1.2.1 : native - OK
                 memory - OK
  Sesame 1.2   : native - FAIL (known bug)
                 memory - OK
  Sesame 1.1.3 : native - OK
                 memory - OK

I've not personally tested Sesame versions earlier than 1.1.3 but they may work.

=head1 COVERAGE

Test coverage results provided by L<Devel::Cover>

 ----------------------------------- ------ ------ ------ ------ ------ ------
 File                                  stmt   bran   cond    sub   time  total
 ----------------------------------- ------ ------ ------ ------ ------ ------
 lib/RDF/Sesame.pm                    100.0    n/a    n/a  100.0    1.1  100.0
 lib/RDF/Sesame/Connection.pm         100.0  100.0  100.0  100.0   20.5  100.0
 lib/RDF/Sesame/Repository.pm          96.7   95.9    n/a  100.0    9.0   96.7
 lib/RDF/Sesame/Response.pm           100.0  100.0    n/a  100.0   46.4  100.0
 lib/RDF/Sesame/TableResult.pm         91.2   82.4    n/a  100.0   23.0   88.7
 Total                                 96.2   92.2  100.0  100.0  100.0   95.4
 ----------------------------------- ------ ------ ------ ------ ------ ------

=head1 DEPENDENCIES
 
A list of all the other modules that this module relies upon, including any
restrictions on versions, and an indication whether these required modules are
part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.
 

=head1 SEE ALSO

=over 4

=item L<RDF::Sesame::Connection>

=item L<RDF::Sesame::Repository>

=item L<RDF::Sesame::TableResult>

=back

=head1 ACKNOWLEDGEMENTS

Southwest Counseling Service for sponsoring the initial development
L<http://swcounseling.org>.

=head1 AUTHOR

Michael Hendricks <michael@ndrix.com>

=head1 LICENSE AND COPYRIGHT
 
Copyright (c) 2005-2006 Michael Hendricks (<michael@ndrix.com>). All rights
reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
